lib/permission.js

```diff
@@ -17,21 +23,164 @@ module.exports = api;
 
 var logger = bedrock.loggers.get('app');
 
-bedrock.events.on('bedrock.start', init);
+bedrock.events.on('bedrock-mongodb.ready', function(callback) {
+  PERMISSIONS = bedrock.config.permission.permissions;
+  async.auto({
+    openCollections: function(callback) {
+      database.openCollections(['roles'], callback);
+    },
+    createIndexes: ['openCollections', function(callback) {
+      database.createIndexes([{
+        collection: 'roles',
+        fields: {id: 1},
+        options: {unique: true, background: false}
+      }], callback);
+    }],
+    addRoles: ['createIndexes', function(callback) {
+      async.eachSeries(
+        bedrock.config.permission.roles, function(role, callback) {
+        role.permanent = true;
+        logger.debug('adding role', role);
+        api.addRole(null, role, function(err) {
+          if(err && database.isDuplicateError(err)) {
+            err = null;
+          }
+          callback(err);
+        });
+      }, callback);
+    }]
+  }, callback);
+});
 
-function init(callback) {
-  // log permissions
-  var permissions = bedrock.config.permission.permissions;
-  Object.keys(permissions).forEach(function(permission) {
-    logger.debug('adding permission', permissions[permission]);
-  });
-  // log roles
-  var roles = bedrock.config.permission.roles;
-  Object.keys(roles).forEach(function(role) {
-    logger.debug('adding role', roles[role]);
-  });
-  callback();
-}
+/**
+ * Returns a role based on the id.
+ *
+ * @param actor the Identity performing the action.
+ * @param id The id of the role.
+ * @param callback(err, role) Called once the operation completes.
+ */
+api.getRole = function(actor, id, callback) {
+  async.waterfall([
+    function(callback) {
+      api.checkPermission(actor, PERMISSIONS.ROLE_ACCESS, callback);
+    },
+    function(callback) {
+      var q = {id: database.hash(id)};
+      database.collections.roles.findOne(q, function(err, result) {
+        if(!result) {
+          return callback(new BedrockError(
+            'Failed to find role.', 'NotFound', {
+            httpStatusCode: 404,
+            public: true
+          }));
+        }
+        callback(err, result.role);
+      });
+    }
+  ], callback);
+};
+
+/**
+ * Returns all roles in the database.
+ *
+ * @param actor the Identity performing the action.
+ * @param callback(err, roles) Called once the operation completes.
+ */
+api.getRoles = function(actor, callback) {
+  async.waterfall([
```

Comment:
I wonder if we could run into a problem here if someone tries to get the roles before mongodb is initialized. Previously, that would work and you'd get the roles from the config just fine. I think we should check to see if our mongodb init handler has run (set a flag) and, if not, use `roles = bedrock.config.permission.roles;` and return early ... which should produce old behavior. Once mongodb has initialized, we can use the code as is here.

That would ensure that if someone tries to do something prior to mongodb loading, permissions will be checked against the permanent roles in the config system. This would be a "just in case" stop gap measure for early module behavior.